File: scripts/generate_vapid.php
#!/usr/bin/env php
<?php
/**
 * Generate VAPID key pair for Web Push (.env).
 *
 * Output format matches web-push / pywebpush (base64url P-256 keys).
 *
 * Usage: php scripts/generate_vapid.php
 */

declare(strict_types=1);

if (PHP_SAPI !== 'cli') {
    fwrite(STDERR, "CLI only.\n");
    exit(1);
}

function b64url(string $bin): string
{
    return rtrim(strtr(base64_encode($bin), '+/', '-_'), '=');
}

$key = openssl_pkey_new([
    'private_key_type' => OPENSSL_KEYTYPE_EC,
    'curve_name'       => 'prime256v1',
]);

if ($key === false) {
    fwrite(STDERR, "OpenSSL failed to generate EC key.\n");
    exit(1);
}

$details = openssl_pkey_get_details($key);
if ($details === false || ($details['type'] ?? 0) !== OPENSSL_KEYTYPE_EC) {
    fwrite(STDERR, "Could not read EC key details.\n");
    exit(1);
}

$d = $details['ec']['d'] ?? '';
$x = $details['ec']['x'] ?? '';
$y = $details['ec']['y'] ?? '';

if ($d === '' || $x === '' || $y === '') {
    fwrite(STDERR, "Could not extract EC coordinates.\n");
    exit(1);
}

$d = str_pad($d, 32, "\x00", STR_PAD_LEFT);
$x = str_pad($x, 32, "\x00", STR_PAD_LEFT);
$y = str_pad($y, 32, "\x00", STR_PAD_LEFT);

$publicKey  = b64url("\x04" . $x . $y);
$privateKey = b64url($d);

$subject = 'mailto:nexalert@yourdomain.com';
$from    = getenv('MAIL_FROM_ADDRESS') ?: '';
if ($from !== '' && str_contains($from, '@') && str_contains($from, '.')) {
    $subject = 'mailto:' . $from;
}

echo "Add to .env:\n\n";
echo "VAPID_PUBLIC_KEY={$publicKey}\n";
echo "VAPID_PRIVATE_KEY={$privateKey}\n";
echo "VAPID_SUBJECT={$subject}\n";

File: scripts/generate_vapid.py
#!/usr/bin/env python3
"""Generate VAPID keys compatible with pywebpush (same format as web-push CLI)."""

from __future__ import annotations

import os
import sys

try:
    from cryptography.hazmat.primitives import serialization
    from py_vapid import Vapid02
    from py_vapid.utils import b64urlencode
except ImportError:
    print("Install: pip install py-vapid", file=sys.stderr)
    sys.exit(1)


def vapid_public_b64url(vapid: Vapid02) -> str:
    raw = vapid.public_key.public_bytes(
        serialization.Encoding.X962,
        serialization.PublicFormat.UncompressedPoint,
    )
    encoded = b64urlencode(raw)
    return encoded.decode() if isinstance(encoded, bytes) else str(encoded)


def vapid_private_b64url(vapid: Vapid02) -> str:
    raw = vapid.private_key.private_numbers().private_value.to_bytes(32, "big")
    encoded = b64urlencode(raw)
    return encoded.decode() if isinstance(encoded, bytes) else str(encoded)


v = Vapid02()
v.generate_keys()

subject = "mailto:nexalert@yourdomain.com"
from_addr = os.environ.get("MAIL_FROM_ADDRESS", "").strip()
if from_addr and "@" in from_addr and "." in from_addr:
    subject = f"mailto:{from_addr}"

print("Add to .env:\n")
print(f"VAPID_PUBLIC_KEY={vapid_public_b64url(v)}")
print(f"VAPID_PRIVATE_KEY={vapid_private_b64url(v)}")
print(f"VAPID_SUBJECT={subject}")
